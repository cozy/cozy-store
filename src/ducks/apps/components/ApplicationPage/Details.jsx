import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'

import { Button } from 'cozy-ui/react/Button'
import Icon from 'cozy-ui/react/Icon'
import Toggle from 'cozy-ui/react/Toggle'

import Maintenance from 'ducks/apps/components/ApplicationPage/Maintenance'

import ReactMarkdownWrapper from 'ducks/components/ReactMarkdownWrapper'
import { getContext, REGISTRY_CHANNELS } from 'ducks/apps'
import getChannel from 'lib/getChannelFromSource'
import { getTranslatedManifestProperty } from 'lib/helpers'
import { isUnderMaintenance } from 'ducks/apps/appStatus'

const isValidUrl = url => {
  if (!url) return null
  return url.match(
    /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w.-]+)+[\w\-._~:/?#[\]@!$&'()*+,;=.]+$/
  )
}

export class Details extends Component {
  constructor(props) {
    super(props)
    const { app } = this.props
    const appChannel = getChannel(app.source)
    this.state = {
      displayBetaChannel: appChannel === REGISTRY_CHANNELS.BETA,
      displayDevChannel: appChannel === REGISTRY_CHANNELS.DEV
    }

    this.toggleChannels = this.toggleChannels.bind(this)
    this.onUpdateChannel = this.onUpdateChannel.bind(this)
  }

  onShowPermissions() {
    const { app, parent } = this.props
    this.props.history.push(`/${parent}/${app.slug}/permissions`)
  }

  async toggleChannels() {
    const { app } = this.props
    // only for installed apps and apps from the registry
    if (!(app.installed && getChannel(app.source))) return
    const stateUpdate = {}
    if (!this.state.displayBetaChannel) {
      stateUpdate.displayBetaChannel = true
    }
    if (!this.state.displayDevChannel) {
      const context = await getContext()
      if (context && context.attributes && context.attributes.debug) {
        stateUpdate.displayDevChannel = true
      }
    }
    this.setState(() => stateUpdate)
  }

  onUpdateChannel(checked, channel) {
    const { app, history, parent } = this.props
    const targetChannel = checked ? channel : REGISTRY_CHANNELS.STABLE
    history.push(`/${parent}/${app.slug}/channel/${targetChannel}`)
  }

  render() {
    const { t, app, description, changes, mobileApps } = this.props
    const { displayBetaChannel, displayDevChannel } = this.state
    const { source, slug, langs, categories, developer, version } = app
    const appChannel = getChannel(source)
    const isBeta = appChannel === REGISTRY_CHANNELS.BETA
    const isDev = appChannel === REGISTRY_CHANNELS.DEV
    const langsInfos = langs && langs.map(l => t(`app_langs.${l}`))
    const categoriesInfos =
      categories &&
      !!categories.length &&
      categories.map(c => t(`app_categories.${c}`))
    const developerName =
      developer && getTranslatedManifestProperty(app, 'developer.name', t)
    const developerUrl =
      developer && getTranslatedManifestProperty(app, 'developer.url', t)
    const shortVersion = version && version.match(/^(\d+\.\d+\.\d+)-.*$/)
    const displayedVersion =
      (shortVersion && shortVersion.length && shortVersion[1]) || version
    return (
      <div className="sto-app-details">
        <div className="sto-app-descriptions">
          {isUnderMaintenance(app) && <Maintenance slug={slug} />}
          {description && (
            <div className="sto-app-description">
              <h3 className="u-title-h3">{t('app_page.description')}</h3>
              <ReactMarkdownWrapper source={description} parseEmoji />
            </div>
          )}
          {changes && (
            <div className="sto-app-changes">
              <h3 className="u-title-h3">{t('app_page.changes')}</h3>
              <ReactMarkdownWrapper source={changes} parseEmoji />
            </div>
          )}
        </div>
        <div className="sto-app-additional-details">
          <h3 className="u-title-h3">{t('app_page.infos.title')}</h3>
          <div className="sto-app-info">
            <div className="sto-app-info-header">
              {t('app_page.infos.categories')}
            </div>
            <div className="sto-app-info-content">
              {categoriesInfos &&
                categoriesInfos.join(t('app_categories.list_separator'))}
            </div>
          </div>
          <div className="sto-app-info">
            <div className="sto-app-info-header">
              {t('app_page.infos.version.title')}
            </div>
            <div className="sto-app-info-content">
              <span onDoubleClick={this.toggleChannels}>
                {displayedVersion || t('app_page.infos.version.unknown')}
              </span>
            </div>
          </div>
          {(displayBetaChannel || displayDevChannel) && (
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.beta')}
              </div>
              <div className="sto-app-info-content">
                <Toggle
                  id={`sto-app-${slug}-beta-toggle`}
                  checked={isBeta}
                  onToggle={e =>
                    this.onUpdateChannel(e, REGISTRY_CHANNELS.BETA)
                  }
                />
              </div>
            </div>
          )}
          {displayDevChannel && (
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.dev')}
              </div>
              <div className="sto-app-info-content">
                <Toggle
                  id={`sto-app-${slug}-dev-toggle`}
                  checked={isDev}
                  onToggle={e => this.onUpdateChannel(e, REGISTRY_CHANNELS.DEV)}
                />
              </div>
            </div>
          )}
          {langsInfos && (
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.langs')}
              </div>
              <div className="sto-app-info-content">
                {langsInfos.join(t('app_langs.list_separator'))}
              </div>
            </div>
          )}
          {mobileApps && !!mobileApps.length && (
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.mobile_app')}
              </div>
              <div className="sto-app-info-content sto-app-info-content--mobile-apps">
                {mobileApps.map(a => {
                  const icon = require(`assets/icons/platforms/icon-${
                    a.type
                  }.svg`)
                  return (
                    <a
                      className="sto-app-info-content-icon"
                      href={isValidUrl(a.url) ? a.url : null}
                      target="_blank"
                      rel="noopener noreferrer"
                      key={a.type}
                    >
                      <Icon icon={icon.default} width="24px" height="24px" />
                    </a>
                  )
                })}
              </div>
            </div>
          )}
          <div>
            <Button
              label={t('app_page.permissions.button.label')}
              className="sto-app-permissions-button"
              onClick={() => this.onShowPermissions()}
              subtle
            />
          </div>
          {developerName && (
            <div>
              <h3 className="u-title-h3">{t('app_page.developer_infos')}</h3>
              <div className="sto-app-developer-infos">
                <span>{developerName}</span>
                {isValidUrl(developerUrl) && (
                  <a
                    href={developerUrl}
                    className="sto-app-developer-link"
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    {developerUrl}
                  </a>
                )}
              </div>
            </div>
          )}
        </div>
      </div>
    )
  }
}

export default withRouter(translate()(Details))
