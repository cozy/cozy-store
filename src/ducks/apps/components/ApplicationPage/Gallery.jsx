import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'
import Modal, { ModalContent } from 'cozy-ui/react/Modal'

export class Gallery extends Component {
  constructor(props) {
    super(props)
    this.state = {
      currentImage: null
    }
    this.onClose = this.onClose.bind(this)
  }

  onClick(imageUrl) {
    this.setState({ currentImage: imageUrl })
  }

  onClose() {
    this.setState({ currentImage: null })
  }

  render() {
    const { t, slug, images } = this.props
    const { currentImage } = this.state
    if (!images) return null
    const imagesParts = images.reduce((acc, image, index) => {
      if (index < 2) {
        if (!acc.big) acc.big = []
        acc.big.push(image)
      } else if (index < 5) {
        if (!acc.small) acc.small = []
        acc.small.push(image)
      }
      return acc
    }, {})
    return (
      <div className="sto-app-images">
        {imagesParts.big && !!imagesParts.big.length && (
          <div className="sto-app-big-images">
            {imagesParts.big.map((image, index) => (
              <img
                className="sto-app-big-image"
                src={image}
                alt={`${slug}-image-${index + 1}`}
                key={`${slug}-image-${index + 1}`}
                tabIndex={0}
                onClick={() => this.onClick(image)}
                onKeyUp={e => (e.keyCode === 13 ? this.onClick(image) : null)}
              />
            ))}
          </div>
        )}
        {imagesParts.small && !!imagesParts.small.length && (
          <div
            className={`sto-app-small-images${
              imagesParts.small.length > 2 ? ' --space-around' : ''
            }`}
          >
            {imagesParts.small.map((image, index) => (
              <img
                className="sto-app-small-image"
                src={image}
                alt={`${slug}-image-${index + 3}`}
                key={`${slug}-image-${index + 3}`}
                tabIndex={0}
                onClick={() => this.onClick(image)}
                onKeyUp={e => (e.keyCode === 13 ? this.onClick(image) : null)}
              />
            ))}
          </div>
        )}
        {currentImage && (
          <Modal
            title={t('app_page.preview')}
            secondaryAction={this.onClose}
            size="xxlarge"
            mobileFullscreen
          >
            <ModalContent>
              <img
                className="sto-app-image-preview"
                src={currentImage}
                alt="image-preview"
              />
            </ModalContent>
          </Modal>
        )}
      </div>
    )
  }
}

export default translate()(Gallery)
