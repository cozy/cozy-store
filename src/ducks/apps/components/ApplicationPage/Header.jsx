import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import compose from 'lodash/flowRight'

import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import withBreakpoints from 'cozy-ui/transpiled/react/helpers/withBreakpoints'
import Intents from 'cozy-interapp'
import { withClient } from 'cozy-client'

import cozySmileIcon from 'assets/icons/icon-cozy-smile.svg'
import AsyncButton from 'ducks/components/AsyncButton'

import { APP_TYPE, getAppIconProps } from 'ducks/apps'
import {
  hasPendingUpdate,
  isUnderMaintenance,
  isInstalledAndNothingToReport
} from 'ducks/apps/appStatus'

export const Header = ({
  t,
  app,
  namePrefix,
  name,
  description,
  parent,
  isInstalling,
  breakpoints = {},
  client
}) => {
  const { slug, installed, type, related, uninstallable } = app
  const { isMobile } = breakpoints
  const isCurrentAppInstalling = isInstalling === slug
  const openApp = link => {
    window.location.assign(link)
  }
  const isKonnector = type === APP_TYPE.KONNECTOR
  const isInstallDisabled = !!isUnderMaintenance(app) || isInstalling
  const isUninstallDisabled = !uninstallable || isCurrentAppInstalling
  return (
    <div className="sto-app-header">
      <div className="sto-app-header-icon">
        <AppIcon app={app} className="sto-app-icon" {...getAppIconProps()} />
      </div>
      <div className="sto-app-header-content">
        <h2 className="sto-app-header-title">
          {namePrefix ? `${namePrefix} ${name}` : name}
        </h2>
        <p className="sto-app-header-description">{description}</p>
        {isInstalledAndNothingToReport(app) && !isCurrentAppInstalling ? (
          isKonnector ? (
            <AsyncButton
              asyncAction={() => {
                const intents = new Intents({ client })
                return intents.redirect('io.cozy.accounts', {
                  konnector: app.slug
                })
              }}
              className="c-btn"
              icon="openwith"
              label={t('app_page.konnector.open')}
            />
          ) : (
            <Button
              onClick={() => openApp(related)}
              className="c-btn"
              icon="openwith"
              label={t('app_page.webapp.open')}
            />
          )
        ) : (
          <Button
            tag={Link}
            to={`/${parent}/${slug}/install`}
            theme="primary"
            extension={isMobile ? 'full' : null}
            disabled={isInstallDisabled}
            onClick={isInstallDisabled ? e => e.preventDefault() : null}
            aria-busy={isCurrentAppInstalling}
            icon={cozySmileIcon}
            label={
              hasPendingUpdate(app)
                ? t('app_page.update')
                : t('app_page.install')
            }
          />
        )}
        {installed && (
          <Button
            tag={Link}
            to={`/${parent}/${slug}/uninstall`}
            theme="secondary"
            extension={isMobile ? 'full' : null}
            className={isMobile ? 'u-mt-1' : null}
            onClick={isUninstallDisabled ? e => e.preventDefault() : null}
            disabled={isUninstallDisabled}
            label={t('app_page.uninstall')}
          />
        )}
      </div>
    </div>
  )
}

export default compose(
  connect(state => ({
    isInstalling: state.apps.isInstalling
  })),
  withBreakpoints(),
  translate(),
  withClient
)(Header)
