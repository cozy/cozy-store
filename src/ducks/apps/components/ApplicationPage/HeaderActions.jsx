import { APP_TYPE, openApp } from '@/ducks/apps'
import {
  hasPendingUpdate,
  isUnderMaintenance,
  isInstalledAndNothingToReport
} from '@/ducks/apps/appStatus'
import AsyncButton from '@/ducks/components/AsyncButton'
import React from 'react'
import { Link, useLocation } from 'react-router-dom'

import { useClient } from 'cozy-client'
import { isFlagshipApp } from 'cozy-device-helper'
import { useWebviewIntent } from 'cozy-intent'
import Button from 'cozy-ui/transpiled/react/Buttons'
import Icon from 'cozy-ui/transpiled/react/Icon'
import CloudIcon from 'cozy-ui/transpiled/react/Icons/Cloud'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

import { handleIntent } from './helpers'

const HeaderActions = ({ app, intentData, parent, isInstalling }) => {
  const client = useClient()
  const { t } = useI18n()
  const { isMobile } = useBreakpoints()

  const isCurrentAppInstalling = isInstalling === app.slug
  const { search } = useLocation()
  const webviewIntent = useWebviewIntent()

  const isKonnector = app.type === APP_TYPE.KONNECTOR
  const isInstallDisabled = !!isUnderMaintenance(app) || isInstalling
  const isUninstallDisabled = !app.uninstallable || isCurrentAppInstalling
  const appOrKonnectorLabel = isKonnector
    ? t('app_page.webapp.open')
    : t('app_page.konnector.open')

  const handleClick = () => {
    openApp(webviewIntent, app)
  }

  const openConnector = async () => {
    if (isFlagshipApp()) {
      return webviewIntent?.call('openApp', app.related, app)
    } else {
      handleIntent(client, intentData, app)
    }
  }

  return (
    <>
      {isInstalledAndNothingToReport(app) && !isCurrentAppInstalling ? (
        isKonnector ? (
          <AsyncButton
            asyncAction={openConnector}
            className="c-btn"
            label={appOrKonnectorLabel}
          />
        ) : (
          <Button
            onClick={handleClick}
            className="c-btn"
            label={appOrKonnectorLabel}
          />
        )
      ) : (
        <Button
          component={Link}
          to={`/${parent}/${app.slug}/install${search}`}
          fullWidth={isMobile ? true : false}
          disabled={isInstallDisabled}
          startIcon={<Icon icon={CloudIcon} />}
          label={
            hasPendingUpdate(app) ? t('app_page.update') : t('app_page.install')
          }
          aria-busy={isCurrentAppInstalling}
          onClick={isInstallDisabled ? e => e.preventDefault() : null}
        />
      )}
      {app.installed && (
        <Button
          className={isMobile ? 'u-mt-1' : null}
          component={Link}
          to={`/${parent}/${app.slug}/uninstall${search}`}
          variant="secondary"
          fullWidth={isMobile ? true : false}
          disabled={isUninstallDisabled}
          label={t('app_page.uninstall')}
          onClick={isUninstallDisabled ? e => e.preventDefault() : null}
        />
      )}
    </>
  )
}

export { HeaderActions }
