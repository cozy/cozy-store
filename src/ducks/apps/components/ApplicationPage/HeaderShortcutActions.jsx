import cozySmileIcon from 'assets/icons/icon-cozy-smile.svg'
import { buildFileFromPathQuery } from 'ducks/queries'
import React, { useState } from 'react'

import { useClient, useFetchShortcut } from 'cozy-client'
import { move } from 'cozy-client/dist/models/file'
import flag from 'cozy-flags'
import Button from 'cozy-ui/transpiled/react/Buttons'
import Icon from 'cozy-ui/transpiled/react/Icon'
import { useAlert } from 'cozy-ui/transpiled/react/providers/Alert'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

const HeaderShortcutActions = ({ app }) => {
  const { isMobile } = useBreakpoints()
  const { t } = useI18n()
  const client = useClient()
  const { shortcutInfos } = useFetchShortcut(client, app.id)
  const { showAlert } = useAlert()

  const [isBusy, setBusy] = useState(false)

  const config = flag('store.alternative-source')

  const handleAdd = async () => {
    try {
      setBusy(true)
      // Install can be disabled
      const shortcutCategory = app.categories[0]
      const categoryPath = config.categories[shortcutCategory]
      const fileQuery = buildFileFromPathQuery(categoryPath)
      const categoryDirectory = await client.fetchQueryAndGetFromState(
        fileQuery
      )
      await move(client, app, categoryDirectory.data[0])
      showAlert({
        message: t('HeaderShortcutActions.addSuccess'),
        severity: 'success'
      })
    } catch {
      showAlert({
        message: t('HeaderShortcutActions.error'),
        severity: 'error'
      })
    } finally {
      setBusy(false)
    }
  }

  const handleRemove = async () => {
    try {
      setBusy(true)
      // Uninstall can be disabled
      const storePath = config.store
      const fileQuery = buildFileFromPathQuery(storePath)
      const storeDirectory = await client.fetchQueryAndGetFromState(fileQuery)
      await move(client, app, storeDirectory.data[0])
      showAlert({
        message: t('HeaderShortcutActions.removeSuccess'),
        severity: 'success'
      })
    } catch {
      showAlert({
        message: t('HeaderShortcutActions.error'),
        severity: 'error'
      })
    } finally {
      setBusy(false)
    }
  }

  const url = shortcutInfos?.data?.attributes?.url || ''

  return (
    <>
      <Button
        href={url}
        label={t('HeaderShortcutActions.open')}
        className="c-btn"
        disabled={isBusy}
      />
      {app.installed ? (
        <Button
          variant="secondary"
          fullWidth={isMobile}
          className={isMobile ? 'u-mt-1' : null}
          onClick={handleRemove}
          label={t('HeaderShortcutActions.remove')}
          busy={isBusy}
        />
      ) : (
        <Button
          variant="secondary"
          fullWidth={isMobile}
          className={isMobile ? 'u-mt-1' : null}
          onClick={handleAdd}
          startIcon={<Icon icon={cozySmileIcon} />}
          label={t('HeaderShortcutActions.add')}
          busy={isBusy}
        />
      )}
    </>
  )
}

export { HeaderShortcutActions }
