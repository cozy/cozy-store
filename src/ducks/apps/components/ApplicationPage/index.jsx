/* global cozy */
import React, { Component } from 'react'
import { Link } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import Button from 'cozy-ui/react/Button'
import FocusTrap from 'focus-trap-react'
import AppIcon from 'cozy-ui/react/AppIcon'

import ApplicationPageLoading from 'ducks/components/ApplicationPageLoading'
import { getTranslatedManifestProperty } from 'lib/helpers'
import Header from './Header'
import Gallery from './Gallery'
import Details from './Details'

import {
  preventBackgroundScroll,
  unpreventBackgroundScroll
} from 'lib/scrollHelpers.js'
import { getAppIconProps } from 'ducks/apps'

const MOBILE_PLATFORMS = ['ios', 'android']
const isMobilePlatform = name => MOBILE_PLATFORMS.includes(name)

const { BarCenter } = cozy.bar

export class ApplicationPage extends Component {
  constructor(props) {
    super(props)
    this.state = { displayBarIcon: false, activeTrap: true }
  }

  componentDidMount() {
    preventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.addEventListener('scroll', this.handleScroll, {
        passive: true
      })
    }
  }

  componentWillUnmount() {
    unpreventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.removeEventListener('scroll', this.handleScroll)
    }
  }

  mountTrap = () => {
    this.setState({ activeTrap: true })
  }

  unmountTrap = () => {
    this.setState({ activeTrap: false })
  }

  handleScroll = () => {
    if (
      !this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop > 100
    ) {
      this.setState(() => ({ displayBarIcon: true }))
    } else if (
      this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop < 100
    ) {
      this.setState(() => ({ displayBarIcon: false }))
    }
  }

  render() {
    const {
      t,
      parent,
      matchRoute,
      isFetching,
      fetchError,
      breakpoints = {},
      pauseFocusTrap,
      getApp,
      redirectTo
    } = this.props
    if (isFetching) return <ApplicationPageLoading />
    const app = getApp(matchRoute)
    if (!app) return redirectTo(`/${parent}`)

    const { displayBarIcon } = this.state
    const { isMobile } = breakpoints
    if (fetchError) {
      return (
        <p className="u-error">
          {t('app_modal.install.message.version_error', {
            message: fetchError.message
          })}
        </p>
      )
    }
    const { icon, slug, iconToLoad } = app
    const appName = getTranslatedManifestProperty(app, 'name', t)
    const namePrefix = getTranslatedManifestProperty(app, 'name_prefix', t)
    const appShortDesc = getTranslatedManifestProperty(
      app,
      'short_description',
      t
    )
    const appLongDesc = getTranslatedManifestProperty(
      app,
      'long_description',
      t
    )
    const appChanges = getTranslatedManifestProperty(app, 'changes', t)
    const mobileApps =
      app.platforms &&
      !!app.platforms.length &&
      app.platforms.reduce((mobilePlatforms, platformInfos) => {
        // force to be lower case
        platformInfos.type = platformInfos.type.toLowerCase()
        if (isMobilePlatform(platformInfos.type)) {
          mobilePlatforms.push(platformInfos)
        }
        return mobilePlatforms
      }, [])
    return (
      <FocusTrap
        focusTrapOptions={{
          onDeactivate: this.unmountTrap,
          clickOutsideDeactivates: true
        }}
        paused={pauseFocusTrap}
      >
        <div className="sto-modal-page-app">
          {isMobile && icon && !iconToLoad && (
            <BarCenter>
              <div className="sto-app-bar">
                <AppIcon
                  app={app}
                  className={`sto-app-bar-icon ${
                    !displayBarIcon ? 'sto-app-bar-icon--hidden' : ''
                  }`}
                  {...getAppIconProps()}
                />
              </div>
            </BarCenter>
          )}
          <div className="sto-app">
            <Button
              icon="back"
              tag={Link}
              to={`${parent}`}
              className="sto-app-back"
              label={t('app_page.back')}
              onClick={this.unmountTrap}
              subtle
            />
            <Header
              app={app}
              namePrefix={namePrefix}
              name={appName}
              description={appShortDesc}
              parent={parent}
            />
            {app.screenshots && !!app.screenshots.length && (
              <Gallery slug={slug} images={app.screenshots} />
            )}
            <Details
              app={app}
              description={appLongDesc}
              changes={appChanges}
              mobileApps={mobileApps}
              parent={parent}
            />
          </div>
        </div>
      </FocusTrap>
    )
  }
}

// translate is needed here for the lang props
export default translate()(withBreakpoints()(ApplicationPage))
