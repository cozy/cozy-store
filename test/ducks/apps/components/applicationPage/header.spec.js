'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import Button from 'cozy-ui/transpiled/react/Button'

import { tMock } from '../../../../jestLib/I18n'
import { Header } from 'ducks/apps/components/ApplicationPage/Header'

import mockApp from '../../_mockPhotosRegistryVersion'
import mockKonnector from '../../_mockPKonnectorTrinlaneRegistryVersion'

window.location.assign = jest.fn()

describe('ApplicationPage header component', () => {
  beforeEach(() => {
    window.location.assign.mockReset()
  })

  it('should be rendered correctly provided app', () => {
    const component = shallow(
      <Header
        t={tMock}
        parent="/myapps"
        namePrefix="Cozy"
        name={mockApp.name}
        description={mockApp.description}
        app={mockApp}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly provided konnector', () => {
    const component = shallow(
      <Header
        t={tMock}
        parent="/myapps"
        app={mockKonnector}
        name={mockKonnector.name}
        description={mockKonnector.description}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly provided app with broken icon and no editor', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.icon = ''
    appProps.editor = ''
    const component = shallow(
      <Header
        t={tMock}
        parent="/myapps"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should handle the click on open installed app button', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.installed = true
    appProps.icon = '<svg></svg>'
    appProps.related = '#'
    const wrapper = shallow(
      <Header
        t={tMock}
        parent="/myapps"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
      />
    )
    // open button is the first one
    wrapper
      .find(Button)
      .at(0)
      .simulate('click')
    expect(window.location.assign.mock.calls.length).toBe(1)
    expect(window.location.assign.mock.calls[0][0]).toBe('#')
  })

  it('should disable the install button if maintenance', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.maintenance = {}
    const wrapper = shallow(
      <Header
        t={tMock}
        parent="/discover"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
      />
    )
    const mockEvent = {
      preventDefault: jest.fn()
    }
    // install button is the only one here
    const installButton = wrapper.find(Button)
    expect(installButton.props().disabled).toBe(true)
    wrapper.find(Button).simulate('click', mockEvent)
    expect(mockEvent.preventDefault.mock.calls.length).toBe(1)
  })

  it('should disable the uninstall button for installed app if not uninstallable', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.installed = true
    appProps.icon = '<svg></svg>'
    appProps.uninstallable = false
    const wrapper = shallow(
      <Header
        t={tMock}
        parent="/discover"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
      />
    )
    const mockEvent = {
      preventDefault: jest.fn()
    }
    // uninstall button is the second one
    const uninstallButton = wrapper.find(Button).at(1)
    expect(uninstallButton.props().disabled).toBe(true)
    uninstallButton.simulate('click', mockEvent)
    expect(mockEvent.preventDefault.mock.calls.length).toBe(1)
  })

  it('should disable the uninstall button an app is installing (updating)', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.installed = true
    const wrapper = shallow(
      <Header
        t={tMock}
        parent="/discover"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
        isInstalling
      />
    )
    const mockEvent = {
      preventDefault: jest.fn()
    }
    // uninstall button is the second one
    const uninstallButton = wrapper.find(Button).at(1)
    expect(uninstallButton.props().disabled).toBe(true)
    uninstallButton.simulate('click', mockEvent)
    expect(mockEvent.preventDefault.mock.calls.length).toBe(1)
  })

  it('should display a placeholder if app icon is still fetching', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.iconToLoad = true
    const component = shallow(
      <Header
        t={tMock}
        parent="/discover"
        app={appProps}
        name={appProps.name}
        description={appProps.description}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
