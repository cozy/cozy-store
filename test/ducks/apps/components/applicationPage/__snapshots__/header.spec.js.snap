// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationPage header component should be rendered correctly provided app 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="photos-icon"
      className="sto-app-icon"
      src="https://mockcozy.cc/registry/photos/3.0.0/icon.svg"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Cozy Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={false}
      onClick={null}
      replace={false}
      to="//myapps/photos/manage"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should be rendered correctly provided app with broken icon and no editor 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <Icon
      className="sto-app-icon--default blurry"
      color="#95999D"
      height="88px"
      icon="test-file-stub"
      spin={false}
      width="88px"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={false}
      onClick={null}
      replace={false}
      to="//myapps/photos/manage"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should be rendered correctly provided konnector 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="konnector-trinlane-icon"
      className="sto-app-icon"
      src="https://mockcozy.cc/registry/konnector-trinlane/0.1.0/icon.svg"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Trinlane
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={false}
      onClick={null}
      replace={false}
      to="//myapps/konnector-trinlane/manage"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should disable the install button if maintenance 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="photos-icon"
      className="sto-app-icon"
      src="https://mockcozy.cc/registry/photos/3.0.0/icon.svg"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={true}
      onClick={[Function]}
      replace={false}
      to="//discover/photos/manage"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should disable the uninstall button for installed app if not uninstallable 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="photos-icon"
      className="sto-app-icon"
      src="<svg></svg>"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <div>
      <DefaultButton
        className="c-btn"
        icon="openwith"
        label="Open"
        onClick={[Function]}
      />
      <Link
        className="c-btn c-btn--danger-outline sto-app-header-uninstall-button"
        disabled={true}
        onClick={[Function]}
        replace={false}
        to="//discover/photos/manage"
      >
        <span>
          Uninstall
        </span>
      </Link>
    </div>
  </div>
</div>
`;

exports[`ApplicationPage header component should display a placeholder if app icon is still fetching 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <Placeholder
      height="8.25rem"
      width="8rem"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={false}
      onClick={null}
      replace={false}
      to="//discover/photos/manage"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Install
      </span>
    </Link>
  </div>
</div>
`;
